doc/docs/spec/typeclasses/index.md

```diff
@@ -0,0 +1,124 @@
+   defined. **Drafting node: need an opt-out mechanism.**
+ * `instance`, used to declare an instance of a typeclass.
+ * `default`, which marks the instance that is used when no instance is explicitly specified.
+ * `with`, which will ultimately be used properly for introducing all kinds of templates, but right now its usage is
+   confined to defining typeclasses, as templates are not a feature to fully land in the minimal viable product.
+
+Similarly to structures, typeclasses can be defined in the named an unnamed forms; the named one is translated to a declaration
+of an entity initialized with the unnamed form of the definition. The grammar is as follows (this duplicates some terms that
+will be defined elsewhere, but not all of them, since quoting the entire grammar of expressions here would be excessive):
+
+```
+named-typeclass-declaration ::= template-introducer `implicit`? `typeclass` identifier
+    `{` typeclass-function* `}` `;`
+typeclass-literal ::= template-introducer `implicit`? `typeclass` `{` function-declaration* `}`
+
+template-introducer ::= `with` `(` parameter (`,` parameter)* `)`
+parameter ::= identifier `:` expression
+
+typeclass-function ::= function-declaration | function-definition
+function-declaration ::= function-signature `;`
+function-definition ::= function-signature block
+function-signature ::= `function` identifier `(` (parameter (`,` parameter)*)? `)` (`->` trailing-type)?
+```
+
+A definition of a default instance is itself a declaration and cannot be used as an expression. Other definitions of
+typeclass instances are expressions and can be used as such.
+
+
+```
+default-instance-definition ::= `default` `instance` identifier `(` expression `)`
+    `{` function-definition* `}` `;`
+instance-literal ::= `instance` identifier `(` expression `)` `{` function-definition* `}`
+```
+
+## Semantics
+
+This section uses names from the following piece of code to define semantics and requirements:
+
+```
+let tc = with (T : type) typeclass {        // 1
+    function fndecl() -> int;               // 2
+
+    function fndef(arg : int) -> int {      // 3
+        return arg;
+    }
+};
+
+default instance tc(int) {                  // 4
+    function fndecl() -> int {              // 5
+        return 0;
+    }
+};
+
+let inst = instance tc(int) {               // 6
+    function fndecl() -> int {              // 7
+        return 1;
+    }
+
+    function fndef(inst_arg : int) -> int { // 8
+        return arg - 1;
+    }
+};
+```
+
+### Basics of typeclasses and instances
+
+ 1. `tc` is a name of a typeclass. The type of `tc` is `typeclass`, which is a builtin type. All typeclasses are of type
+    `typeclass`.
+ 2. `inst` is a name of a typeclass instance. The type of `inst` is `tc`.
+ 3. The type of the unnamed default instance is also `tc`.
+ 4. Values of type that is a typeclass can be used as a type. **Drafting note: this is somewhat unimportant for the time
```

Comment:
This is an important point. I think it's also important to have another point saying that typeclasses can be stored as members, or treated no differently than *something else you have specified to which you can refer.*

Reply:
Your comment is probably a good proof that my sentence here is very confusing. I don't mean values of type `tc` (in this example); I mean values of type `tc(int)`. I should try to make this clearer, and also write down what you said as a separate point.

Reply:
I put this comment on this line only because it seemed like a related place to put it. I wasn't thinking that `tc` alone could be used as a type, but I do see what you mean about the wording. Either way, both are valid points.